Added time traveling use case tutorial (#886)
EventStore/documentation: docs/getting-started/use-cases/time-travel/tutorial-3.md

---
title: Part 3 - Project Events to Sales Report Read Model
---

# Part 3: Project Events to Sales Report Read Model

> Now that KurrentDB is initialized with a few hundred order events, you will learn how to construct a sales report by projecting these events into a JSON file read model, and then rendering it as a sales report on a web page.

Reviewer: ```suggestion
Now that KurrentDB is initialized with a few hundred order events, you will project the events into a JSON file-read model and render it as a sales report on a web page.
```

### Introducing the Month End Sales Report

> The month end sales report shows the monthly total revenue generated from orders placed in the e-commerce business’s stores worldwide. Sales data is organized by product category and by the region where each order was placed.

Reviewer: ```suggestion
The month-end sales report summarizes total revenue from orders placed across all e-commerce stores worldwide. Sales figures are broken down by product category and region.

The report also includes the monthly sales target to support forecasting and performance management.
```

   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ `regions` | Maps regions to their sales data (e.g. Asia, Europe, North-America).                                     |
   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ *region* | Contains sales metrics for the region.                     |
   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ `dailySales`         | Sales for the reported day.           |
   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ `monthEndSalesTarget`| The sales target for the month.                |
   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ `totalMonthlySales`  | The total sales for the month.                 |
   | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└─ `targetHitRate`      | Ratio of actual sales to target.                |
   
   :::

   The structure of `report-read-model.json` is intentionally **denormalized**. This means the data is organized and pre-aggregated for direct use, in this case, by the web page.

   ::: info What Does "Denormalized" Mean?

>    A denormalized read model stores related and pre-aggregated data together, shaped for fast querying and display—such as totals by category and region. Unlike normalized databases, which separate data into many related tables, a denormalized model puts everything needed for a report or UI in one place. This avoids complex joins or calculations at query time and makes it easy for the front end to use the data as-is.

Reviewer: ```suggestion
  A denormalized read model stores related and pre-aggregated data together, such as sales totals by category and region, in a format optimized for fast querying and display. Unlike normalized databases, which separate data into multiple related tables, a denormalized model keeps all the information needed for a report or user interface in one place. This approach eliminates the need for complex joins or calculations at query time, allowing front-end applications to use the data directly.

Denormalization is especially common for read models generated from KurrentDB events, as it enables applications to efficiently present and analyze event data without requiring additional processing at runtime.
```

   }
   ```

   The read model is retrieved by reading the JSON file from a location specified by an environment variable. If the file exists and can be read successfully, it is deserialized into a `ReportReadModel` object. If the file doesn't exist or deserialization fails, a new empty `ReportReadModel` object is created instead.

3. Locate and examine the code that retrieves the last checkpoint:

   ```cs
   // ---------------------------------------------------------- //
   // Retrieve the last checkpoint position from JSON Read Model //
   // ---------------------------------------------------------- //

   var streamPosition = hasExistingReadModel                                // Check if the checkpoint exists..
      ? FromStream.After(StreamPosition.FromInt64(readModel.Checkpoint))    // if so, subscribe from stream after checkpoint..
      : FromStream.Start;                                                   // otherwise, subscribe from the start of the stream
   ```

   The checkpoint is retrieved from the deserialized read model. If no read model/checkpoint is found or it is the first time the application is executed, we can retrieve the default start position.

>    ::: info Understanding Checkpoint

Reviewer: ```suggestion
   ::: info Understanding Checkpoints
```

   A projection often uses a checkpoint to recover the position of the last processed event. This way, when an application unexpectedly crashes mid-process, the projection does not have to process all the previously processed events.
   :::

   ::: info Storing Checkpoints for Read Models Stored as File
      Checkpoints for read models that are stored as files can often be conveniently saved within the file along with the data structure. This way both the checkpoint and read model can be atomically saved together:

      ```json
      {
         "checkpoint": 42,     // <--- CHECKPOINT
         "salesReports": {     // <--- READ MODEL
            "2025-01-31": {
               "categories": {
                  "Electronics": {
                     "regions": {
                        "Asia": {
                           "dailySales": 5200.25,
                           "monthEndSalesTarget": 6000.00,
                           "totalMonthlySales": 60200.90,
                           "targetHitRate": 86.71
                        },
                        // ...
      ```
   :::

4. Locate and examine the code that subscribes to stream:

>    ```cs

Reviewer: The previous images contain the commented out "title" of the code snippet.

Author: Hey Mark, sorry didn't catch what you mean. Can you elaborate?

   await using var subscription = kurrentdb.SubscribeToStream(              // Subscribe to events..
      "$et-order-placed",                                                  // from the order placed event type system projection..        
      streamPosition,                                                      // from this position..
      true);                                                               // with linked events automatically resolved (required for system projections)

   ```

   A subscription is created that subscribes to events from the `$et-order-placed` stream. The subscription will only retrieve events starting from `streamPosition` in the stream (i.e., the checkpoint retrieved from the previous step).

   ::: info Different Types of Subscriptions
   This sample uses catch-up subscriptions to subscribe to events. You can also use persistent subscriptions or connectors to achieve a similar result. 

   For more information about catch-up subscriptions, [click here](/clients/grpc/subscriptions.html).
   For more information about persistent subscriptions, [click here](/clients/grpc/persistent-subscriptions.html). 

>    For more information about connectors, [click here](/server/v24.10/features/connectors/)

Reviewer: ```suggestion
   For more information about connectors, [click here](/server/v24.10/features/connectors/).
```

   :::

5. Locate and examine the code that processes each event:

   ```cs
   await foreach (var message in subscription.Messages)                     // Iterate through the messages in the subscription
   {
      if (message is not StreamMessage.Event(var e)) continue;             // Skip this message if it is not an event

      if (EventEncoder.Decode(e.Event.Data, "order-placed")                // Try to deserialize the event to an OrderPlaced event
               is not OrderPlaced orderPlaced) continue;                    // Skip this message if it is not an OrderPlaced event

      ReportProjection.ReportProjection.ProjectOrderToReadModel(           // Project the event to the read model
         orderPlaced, readModel);

      readModel.Checkpoint = e.OriginalEventNumber.ToInt64();              // Set the read model checkpoint to the event number of the event we just read

      SaveReadModel(readModel, readModelPath);                             // Save the read model to the JSON file

      Console.WriteLine($"Projected event " +
                        $"#{e.OriginalEventNumber.ToInt64()} " +
                        $"{e.Event.EventType}");
   }
   ```

>    For each event, loop:

Reviewer: ```suggestion
   For each event, the previous code:
```

   - Updates the read model with each order via `ProjectOrderToReadModel`
   - Tracks progress by updating the checkpoint in the read model after each event
   - Persists changes by saving the updated model to JSON
   - Runs continuously to handle the event stream

   ::: warning Updating Read Model and Checkpoints Atomically
   To ensure atomicity and consistency, the updates to the read model and checkpoint should be committed within the same transaction. This guarantees that the update to both will succeed or fail together, preventing data inconsistencies like outdated read models or incorrect checkpoint positions. It also simplifies error recovery and ensures the system remains in sync.
   :::

   ::: info Exactly-once processing with Catch-up Subscription
   This implementation ensures exactly-once processing by using KurrentDB for reliable persistence, idempotent projection logic, and transactional updates. The read model and checkpoint are updated atomically, preventing duplicates or inconsistencies, unlike traditional message brokers that rely on at-least-once or at-most-once delivery.
   :::

## Step 7: Examine the Report Projection Logic

> The report projection logic transforms order event into the month end sales reports. It is located in `./ReportProject/ReportProjection.cs`, which:

Reviewer: ```suggestion
In this step, you will explore the report projection logic that transforms the order events into the month-end sales reports. The code located in `./ReportProject/ReportProjection.cs`:
```

- Takes an OrderPlaced event and a ReportReadModel as inputs
- Extracts key information like order date and geographic region
- Updates two sales metrics in the report for each product category and region:
   - Monthly sales totals
   - Daily sales

> In this step, you will explore how the logic achieves this.

Reviewer: ```suggestion
```

Reviewer: delete this sentence - its included in the opening sentence now.

   
   :::

   The structure of `report-read-model.json` is intentionally **denormalized**. This means the data is organized and pre-built for direct use, in this case, by the web page.

   ::: info What Does "Denormalized" Mean?
   A denormalized read model stores related and pre-computed data together, such as sales totals by category and region, in a format optimized for fast querying and display. Unlike normalized data structures, which separate data into multiple related models, a denormalized model keeps all the information needed for a report or user interface in one place. This approach eliminates the need for complex joins or calculations at query time, allowing front-end applications to use the data directly.

   Denormalization is especially common for read models generated from KurrentDB events, as it enables applications to efficiently present and analyze event data without requiring additional processing at runtime.

   Denormalization is a common for read models generated from KurrentDB events, since it allows applications to efficiently present and analyze event data without extra processing.

>    :::

Reviewer: This doesn't seem to be rendering correctly.  The ":::" is listed after the final word (processing).

        var orderRegion = orderPlaced.Store!.geographicRegion!;                 // Get the order region from the store object

        var finalDayOfTheMonth =                                                // Get the last day of the month for the order date
            new DateOnly(orderDate.Year, orderDate.Month, 
                DateTime.DaysInMonth(orderDate.Year, orderDate.Month));

        ProjectToMonthEndReport();                                              // Project the order to the month-end report

        //...
   }
   ```

   `ProjectOrderToReadModel()` is the main method of the projection which takes the `OrderPlaced` event and updates the `ReportReadModel` that is passed in.

   The method prepares several variables for the projection and calls `ProjectToMonthEndReport()`.

3. Locate and examine `ProjectToMonthEndReport()`:

   ```cs
   void ProjectToMonthEndReport()                                          // Project the order to the monthly sales for the last day of the month
   {
      ProjectToMonthlySales(finalDayOfTheMonth);                           // Project to monthly sales for the last day of the month

>       if (orderDate == finalDayOfTheMonth)                                 // If the order date is the last day of the month{

Reviewer: open bracket at end of note
